### PR TITLE
GBA: Fix wave RAM write/read behaviour

### PR DESCRIPTION
**Pull request by AntonioND**
Check the individual commit messages for more details.

### CODE REVIEW

**Comment by AntonioND** [comment]
I have created this test ROM to make sure that my changes are correct. It passes in no$gba, but fails in mgba. It almost passes on mgba after the changes in this PR, but it looks like the default waveform is still incorrect in one case in mgba.

[wave_ram_test.zip](https://github.com/mgba-emu/mgba/files/6178087/wave_ram_test.zip)

A successful run of this test should just print the tests as they run (no "Failed" anywhere).

File: src/gba/audio.c
Comment (endrift) [inline src/gba/audio.c:165]:
```suggestion
	if (!audio->enable) {
		return;
	}
```

File: src/gba/audio.c
Comment (endrift) [inline src/gba/audio.c:235]:
```suggestion
	if (!audio->enable) {
		bank = 1;
	}
```

Comment (endrift) [reply src/gba/audio.c:235]: Further, it the comment can be a bit confusing given the fact that it says "0" but assigns 1. If you put the negation after the block it'll probably be clearer.

**Review by endrift — CHANGES_REQUESTED** [review]
Some style nits mostly

File: src/gba/audio.c
Comment (AntonioND) [reply src/gba/audio.c]: Done.

File: src/gba/audio.c
Comment (AntonioND) [reply src/gba/audio.c]: I see what you mean. I've reworded the comment a bit, what do you think now?

Also, I forgot to commit the same check in the function to write to wave RAM, I've added the change to the commit now.

**Review by endrift — APPROVED** [review]

**Comment by endrift** [comment]
I want to test this on some games that make heavy use of channel 3 before I merge it but otherwise it should be fine.

File: src/gba/audio.c
Comment (endrift) [inline src/gba/audio.c:166]: Actually, I realize that this check needs to be done _outside_ of the function since stuff gets written into the I/O memory buffer outside of this function. That probably is similar for all of the other audio writes...there's actually a TODO about this in GBIORead.

Comment (AntonioND) [reply src/gba/audio.c:166]: Well, according to GBATEK:

> While Bit 7 is cleared, both PSG and FIFO sounds are disabled, and all PSG registers at 4000060h..4000081h are reset to zero (and must be re-initialized after re-enabling sound). However, registers 4000082h and 4000088h are kept read/write-able (of which, 4000082h has no function when sound is off, whilst 4000088h does work even when sound is off).

Comment (endrift) [reply src/gba/audio.c:166]: It should be done as a separate commit at least, but same PR might be fine.

File: src/gba/io.c
Comment (AntonioND) [inline src/gba/io.c:358]: I've had to modify the flow of this function to avoid having checks in each case of each sound register. If you check this branch locally with gitk and tick the "ignore whitespace changes" you can see that the actual changes are minimal.

**Comment by AntonioND** [comment]
This is another test ROM that also checks the initial state of the wave RAM:

[wave_ram.zip](https://github.com/mgba-emu/mgba/files/6186335/wave_ram.zip)

Unfortunately, it doesn't use libgba, it uses my library (https://github.com/AntonioND/ugba), so I've included source, ROM, and reference image. After all, I've detected this problem in mGBA because I was developing tests for my library and using mGBA for quick testing.

File: src/gba/io.c
Comment (endrift) [inline src/gba/io.c:351]: Merge these two conditionals.

File: src/gba/io.c
Comment (endrift) [inline src/gba/io.c:346]: Merge these two lines.

**Review by endrift — CHANGES_REQUESTED** [review]
I probably wouldn't have changed the control flow that way, but it's not actually bad to do it that way. That said, there are some extra bits that should be merged in the process.

File: src/gba/audio.c
Comment (endrift) [inline src/gba/audio.c:60]: This should probably be moved to `GBAudioReset` and the pattern initialization be moved to `GBSkipBIOS` then.

**Review by endrift — CHANGES_REQUESTED** [review]

File: src/gb/gb.c
Comment (endrift) [inline src/gb/gb.c:569]: I'm not sure this will compile everywhere since you're putting a variable declaration in the middle of a `switch` statement. That makes some compilers very unhappy.

**Review by endrift — APPROVED** [review]

File: src/gb/gb.c
Comment (AntonioND) [reply src/gb/gb.c:569]: I can put it inside the for, but I saw that this was the style of this file.

Comment (AntonioND) [reply src/gb/gb.c:569]: Also, I'm not sure about whether to initialize the wave RAM to the pattern always, or just in GCB and AGB.

Comment (endrift) [reply src/gb/gb.c:569]: No, that's worse. You might need to declare it _outside_ of the switch, or put a block inside the case.

As for CGB and AGB vs always, I'd err towards always since that's the old behavior.

Comment (AntonioND) [reply src/gb/gb.c:569]: In that case, I can just leave the loop outside of the switch for now.

File: src/gb/gb.c
Comment (AntonioND) [inline src/gb/gb.c:571]: Actually, this is wrong, this means that when the emulator starts in DMG mode with the boot ROM enabled the wave RAM will be 0 instead of having the pattern.

Comment (endrift) [reply src/gb/gb.c:571]: What is the default pattern on the DMG?

Comment (AntonioND) [reply src/gb/gb.c:571]: https://github.com/LIJI32/SameBoy/blob/72cb391612863a43cb3c876b1d210534a54a4020/Core/gb.c#L1286-L1315

Random, as the TODO comment said.

Comment (endrift) [reply src/gb/gb.c:571]: Is that a problem? You're right that it's different from the old behavior though.

File: src/gba/io.c
Comment (endrift) [inline src/gba/io.c:349]: These extra parens aren't needed.

File: src/gb/gb.c
Comment (AntonioND) [reply src/gb/gb.c:571]: So the right behaviour would be that ``GBReset()`` always initializes the wave RAM to random values, and ``GBSkipBIOS()`` initializes the GBC/GBA values. However, that changes the behaviour, so I think that in this patch I should just add this loop to ``GBReset()`` unconditionally.

Comment (endrift) [reply src/gb/gb.c:571]: Doesn't that just revert what you'd done prior?

Comment (AntonioND) [reply src/gb/gb.c:571]: Well, it's a bit of a revert, but the idea is that I don't want to deal with the DMG randomness of the wave RAM, that's a topic that I don't understand well enough. I guess that the result I'm looking for right now is that the wave RAM has the right values (zero) in the case of GBA in GBA mode, and it behaves the same way as it did before the PR for all other cases.

Comment (endrift) [reply src/gb/gb.c:571]: I suppose that's fair for right now. I'd like to merge this before the release since it makes subtle API changes, so you have like two days to finish the PR :P

Comment (AntonioND) [reply src/gb/gb.c]: 😅

Comment (AntonioND) [reply src/gb/gb.c]: Ok, it should be fine now.

Comment (endrift) [reply src/gb/gb.c]: Honestly at this point the only perceptible difference from the third commit is that...now the memory is always zeroed on AGB (excluding CGB-on-AGB), and there's a more comprehensive comment...but it's in the wrong file now. I'm not sure this helps at all heh.

Comment (AntonioND) [reply src/gb/gb.c]: Well, the initial state of the wave RAM has some effect in at least the game R-Type: https://gbdev.gg8.se/wiki/articles/Gameboy_sound_hardware#Power_Control I don't know if there is any GBA game affected by this, though. However, I was testing my code to use the PSG channel 3 and noticed it worked on mGBA but not on hardware, so it's not a stretch to assume that someone else may find the same problem. 

Considering that most people would just assume that it behaves the same way in GBC mode or GBA mode, initializing it to 0 in the GBA is something worth doing. It is a significant difference from the random or square waves of GB mode that, at least, would be heard.

Regarding having it in the wrong files, it depends on how you look at it. The way I see it, it is the responsibility of ``GBSkipBIOS()`` to initialize the pattern in GBC and GBA (GBC mode), not of ``GBAudioReset()`` because it's shared between all GB and GBA modes.

As I said in a previous comment, the right behaviour would be that ``GBReset()`` always initializes the wave RAM to random values (because that's probably what happens, at least on DMG, maybe it changed in GBC but it doesn't make any difference), and ``GBSkipBIOS()`` initializes the GBC/GBA values. However, that would require to randomize the wave RAM in a realistic way.

Comment (AntonioND) [reply src/gb/gb.c]: However, if this is going to be that controversial, I can get rid of the third patch of this PR and push it as a different PR.

Comment (endrift) [reply src/gb/gb.c]: I think that would be a good idea. Plus that one doesn't change API so it can be merged after release if needed.

Comment (AntonioND) [reply src/gb/gb.c]: Ok, removed.

**Pull request merged**
